### PR TITLE
31-bit shim library for z/OS

### PR DESCRIPTION
**Pull request by joransiu**
This PR adds two new modules:

1. An `include31` modules that produces copies jni.h, jni_convert.h and jniport.h.  For jni.h, the m4 has been updated to remap any object reference types, jclass, jfieldID, jmethodID to be 64-bit... as we want to use the 64-bit handles directly passed by the JVM.   

2. A `j9vm31` 31-bit standalone shim library to support the JavaInvokeInterface and JavaNativeInterface APIs that allow
z/OS 31-bit non-XPLINK native code to interoperate with 64-bit JVM on z/OS.

Changes include:
- the CEL4RO64 utility functions to facilitate AMODE64 calls.  Specifically, getting the call arguments into the correct 64-bit representation, and extracting the appropriate return types.  One thing to note is that float/double parameters and return value handling currently invokes empty assembler functions in order to set or extract the FPR0 value out -- by leveraging the system linkage convention to get the compiler to do the right thing.  Ideally, this would have been done with a single inline asm, but alas. 
- The majority of changes are shim functions to redirect to corresponding 64-bit functions for the JavaInvokeInterface and JavaNativeInterface APIs.
    - Majority of the APIs are invoking the 64 counterparts directly.   For Call*, Array, NewObject*, and String* APIs, they will be calling the 31-bit customized version being introduced in https://github.com/eclipse/openj9/pull/12074.   RegisterNatives and JNI_CreateJavaVM need to hightag any 31-bit native functions being passed.  Some of the Invoke APIs require extra handling to zero out upper-32-bits padding.  
    - Organization wise, the shim functions are broken up similar to how they are represented in `vm` directory.
- JavaVM31 and JNIEnv31 structs for 31-bit AMODE native code to interact with JavaInvokeInterface and JavaNativeInterface APIs.  There are special handling of these, but the current LE limitation only supports a single TCB, so some of the additional logic would be exercised.

Net result of this change is a new include31 directory with 3 JNI header files, and a new libjvm31.so / x library and side-deck produced.  Although independent, the functionality in this PR is dependent on changes introduced in https://github.com/eclipse/openj9/pull/12074.

### CODE REVIEW

**Comment by joransiu** [comment]
@gacholio / @pshipton, do you mind reviewing these changes?  Thanks!

One thing to note, I haven't added a CMake setup to the new j9vm31 module.   As you see, there are some compile/link options exclusions that I'm not quite sure how to replicate yet.

File: runtime/j9vm31/jnicgen.cpp [hunk 1]
```diff
@@ -0,0 +1,1520 @@
...
+
+jobject JNICALL
+CallObjectMethod(JNIEnv *env, jobject obj, jmethodID methodID, ...)
+{
+	va_list_64 parms;
+	long long value = *((int *)((char*)&(methodID) + sizeof(jmethodID)));
+	(void)( (parms)[0] = 0, (parms)[1] =  (char *)&(methodID), (parms)[2] = 0, (parms)[3] = (parms)[1] + sizeof(jmethodID) );
+
+	jobject returnValue = CallObjectMethodV_64(env, obj, methodID, parms);
+	(void)( (parms)[0] = 0, (parms)[1] =  (char *)0, (parms)[2] = 0, (parms)[3] = (char *)0);
```
Comment (pshipton) [inline runtime/j9vm31/jnicgen.cpp:45]: What's this for?

Comment (joransiu) [reply runtime/j9vm31/jnicgen.cpp:45]: The last line is not necessary.  I had expanded the va_list macro and updated to 64-bit representation, but at this stage after the call, parms is no longer needed.

File: runtime/j9vm31/j9cel4ro64.c [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+		controlBlock->argumentsOffset = controlBlock->functionOffset;
+	} else {
+		controlBlock->argumentsOffset = controlBlock->functionOffset + functionNameLength + sizeof(functionBlock->length);
+	}
+
+	/* For DLL load operations, we need a module name specified. */
+	moduleBlock = (J9_CEL4RO64_module *)((char *)(controlBlock) + controlBlock->moduleOffset);
+	infoBlock->ro64Module = moduleBlock;
+	moduleBlock->length = moduleNameLength;
+	strncpy(moduleBlock->moduleName, moduleName, moduleNameLength);
```
[the remark this answers is not in the document]
Comment (joransiu) [reply runtime/j9vm31/j9cel4ro64.c:300]: That is a good suggestion.  I'll track and make this change in a future PR (and do it to similar code for libffi and OMR too).

File: runtime/j9vm31/jnicgen.cpp (hunk 1 above)
Comment (joransiu) [reply runtime/j9vm31/jnicgen.cpp]: Cleaned up under: 6ae3a89

File: runtime/j9vm31/j9cel4ro64.c [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+	}
+}
+
+uint32_t
+j9_cel4ro64_call_function(
+	uint64_t functionDescriptor, J9_CEL4RO64_ArgType *argTypes, uint64_t *argValues, uint32_t numArgs,
+	J9_CEL4RO64_ArgType returnType, void *returnStorage)
+{
+	uint32_t retcode = J9_CEL4RO64_RETCODE_OK;
+	uint32_t argumentAreaInBytes = (0 == numArgs) ? 0 : (8 * numArgs + sizeof(uint32_t));  /* 8 byte slots + 4 byte length member. */
```
[the remark this answers is not in the document]
Comment (joransiu) [reply runtime/j9vm31/j9cel4ro64.c]: Updated to use sizeof(uint64_t) in 7facae1.  I'll squash my commits at the end.

**Comment by pshipton** [comment]
@gacholio this is also ready if you don't have any concerns. Although there should be some squashing done first.

**Comment by joransiu** [comment]
I've just squashed the two commits and rebased, after seeing two unrelated commits show up in the PR.

**Comment by gacholio** [comment]
I'll take a look tomorrow.

**Review by gacholio — APPROVED** [review]

**Comment by gacholio** [comment]
jenkins copyright check

**Comment by gacholio** [comment]
jenkins copyright check

**Comment by gacholio** [comment]
jenkins compile win jdk8

**Pull request merged**